node/lib/util/sequencer_state.js

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2018, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert  = require("chai").assert;
+const deeper   = require("deeper");
+
+const TYPE = {
+    CHERRY_PICK: "CHERRY_PICK",
+    MERGE: "MERGE",
+    REBASE: "REBASE",
+};
+
+/**
+ * This module defines the `SequencerState` value-semantic type.
+ */
+
+/**
+ * @class CommitAndRef
+ *
+ * This class describes a commit and optionally the ref it came from.
+ */
+class CommitAndRef {
+    /**
+     * Create a new `CommitAndRef` object.
+     *
+     * @param {String}      sha
+     * @param {String|null} ref
+     */
+    constructor(sha, ref) {
+        assert.isString(sha);
+        if (null !== ref) {
+            assert.isString(ref);
+        }
+        this.d_sha = sha;
+        this.d_ref = ref;
+
+        Object.freeze(this);
+    }
+
+    /**
+     * @property {String} sha  the unique identifier for this commit
+     */
+    get sha() {
+        return this.d_sha;
+    }
+
+    /**
+     * @property {String|null} ref
+     *
+     * If the commit was referenced by a ref, this is its name.
+     */
+    get ref() {
+        return this.d_ref;
+    }
+
+    /**
+     * Return true if the specified `rhs` represents the same value as this
+     * object.  Two `CommitAndRef` objects represet the same value if they have
+     * the same `sha` and `ref` properties.
+     *
+     * @param {CommitAndRef} rhs
+     * @return {Bool}
+     */
+    equal(rhs) {
```

Comment:
I think the types they're referring to in that doc (e.g., `Hash`) are a library that's not part of standard JS, which, unfortunately doesn't seem to have standard operators that I've seen for equality testing and what not. In any case, I made an `equal`, not an `equals`.

Either way, I've been using `equal` in other classes. If this code is OK otherwise, I'd be happy to go back and revisit across the board later.